### PR TITLE
Service management

### PR DESCRIPTION
**Pull request by Buztas**
* Appointments CRUD
* Time slots
* Services CRUD
* Categories CRUD
* Middleware for businessId checking

### CODE REVIEW

File: Shopipy.ApiService/Middleware/BusinessExistsMiddleware.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
+using Microsoft.AspNetCore.Http;
```
[the remark this answers is not in the document]
Comment (Buztas) [reply Shopipy.ApiService/Middleware/BusinessExistsMiddleware.cs:1]: I mean, we can just have this middleware, so there wouldn't be repetition idk, we dont lose anything.

File: Shopipy.AppointmentManagement/Controllers/AppointmentController.cs [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    private readonly AppointmentService _appointmentService;
+    private readonly IMapper _mapper;
+
+    public AppointmentController(AppointmentService appointmentService, IMapper mapper)
+    {
+        _appointmentService = appointmentService;
+        _mapper = mapper;
+    }
+    
+    [HttpGet("{businessId}/employees/available")]
```
Comment (piatrashkakanstantinass) [inline Shopipy.AppointmentManagement/Controllers/AppointmentController.cs:23]: Route prefix is duplicated, extract it to the top of controller.

File: Shopipy.AppointmentManagement/Controllers/AppointmentController.cs [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+namespace AppointmentManagement.Controllers;
+
+[ApiController]
+[Route("[controller]")]
+public class AppointmentController : ControllerBase
+{
+    private readonly AppointmentService _appointmentService;
+    private readonly IMapper _mapper;
+
+    public AppointmentController(AppointmentService appointmentService, IMapper mapper)
```
Comment (piatrashkakanstantinass) [inline Shopipy.AppointmentManagement/Controllers/AppointmentController.cs:17]: Tip: primary constructors

Comment (Buztas) [reply Shopipy.AppointmentManagement/Controllers/AppointmentController.cs:17]: Will look into it

File: Shopipy.AppointmentManagement/Properties/launchSettings.json [hunk 4]
```diff
@@ -0,0 +1,41 @@
+﻿{
```
Comment (piatrashkakanstantinass) [inline Shopipy.AppointmentManagement/Properties/launchSettings.json:1]: What are you launching here?

File: Shopipy.AppointmentManagement/Services/AppointmentService.cs [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    private readonly IGenericRepository<Appointment> _appointmentRepository;
+    private readonly IGenericRepository<Service> _serviceRepository;
+    private readonly IGenericRepository<User> _userRepository;
+
+    public AppointmentService(IGenericRepository<Appointment> appointmentRepository,
+        IGenericRepository<Service> serviceRepository, IGenericRepository<User> userRepository)
+    {
+        _appointmentRepository = appointmentRepository;
+        _serviceRepository = serviceRepository;
+        _userRepository = userRepository;
```
Comment (piatrashkakanstantinass) [inline Shopipy.AppointmentManagement/Services/AppointmentService.cs:19]: Do not access userRepository directly, use userService (PR not merged yet :( )

File: Shopipy.AppointmentManagement/Services/AppointmentService.cs [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+public class AppointmentService
+{
+    private readonly IGenericRepository<Appointment> _appointmentRepository;
+    private readonly IGenericRepository<Service> _serviceRepository;
+    private readonly IGenericRepository<User> _userRepository;
+
+    public AppointmentService(IGenericRepository<Appointment> appointmentRepository,
+        IGenericRepository<Service> serviceRepository, IGenericRepository<User> userRepository)
+    {
+        _appointmentRepository = appointmentRepository;
```
Comment (piatrashkakanstantinass) [inline Shopipy.AppointmentManagement/Services/AppointmentService.cs:17]: I do not see much point in splitting Appointments and Services into separate projects (separate project = candidate for microservice, appointments are too coupled to services)

File: Shopipy.CategoryManagement/Controllers/CategoryController.cs [hunk 7]
```diff
@@ -0,0 +1,78 @@
+using AutoMapper;
+using Microsoft.AspNetCore.Mvc;
+using Shopipy.CategoryManagement.DTOs;
+using Shopipy.CategoryManagement.Services;
+using Shopipy.Persistence.Models;
+
+namespace Shopipy.CategoryManagement.Controllers;
+
+[ApiController]
+[Route("[controller]")]
```
Comment (piatrashkakanstantinass) [inline Shopipy.CategoryManagement/Controllers/CategoryController.cs:10]: See comment regard route prefix

File: Shopipy.CategoryManagement/Properties/launchSettings.json [hunk 8]
```diff
@@ -0,0 +1,41 @@
+﻿{
```
Comment (piatrashkakanstantinass) [inline Shopipy.CategoryManagement/Properties/launchSettings.json:1]: launching what?

File: Shopipy.ServiceManagement/Properties/launchSettings.json [hunk 9]
```diff
@@ -0,0 +1,41 @@
+﻿{
```
Comment (piatrashkakanstantinass) [inline Shopipy.ServiceManagement/Properties/launchSettings.json:1]: that's quite some launching going on

File: Shopipy.ServiceManagement/Shopipy.ServiceManagement.csproj [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+        <TargetFramework>net8.0</TargetFramework>
+        <Nullable>enable</Nullable>
+        <ImplicitUsings>enable</ImplicitUsings>
+        <RestorePackagesWithLockFile>true</RestorePackagesWithLockFile>
+        <DisableImplicitNuGetFallbackFolder>true</DisableImplicitNuGetFallbackFolder>
+    </PropertyGroup>
+
+    <ItemGroup>
+        <ProjectReference Include="..\Shopipy.Persistence\Shopipy.Persistence.csproj" />
+        <ProjectReference Include="..\Shopipy.CategoryManagement\Shopipy.CategoryManagement.csproj" />
```
Comment (piatrashkakanstantinass) [inline Shopipy.ServiceManagement/Shopipy.ServiceManagement.csproj:14]: I would rather put Category management service interface to shared project than doing direct import

File: Shopipy.ServiceManagement/Controllers/ServiceController.cs [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    [HttpGet("{businessId}/services/category/{categoryId}")]
+    public async Task<IActionResult> GetServicesByCategory(int businessId, int categoryId)
+    {
+        var category = await _categoryService.GetCategoryByIdAsync(categoryId);
+        if (category == null)
+        {
+            return NotFound("Category does not exist");
+        }
```
Comment (piatrashkakanstantinass) [inline Shopipy.ServiceManagement/Controllers/ServiceController.cs:53]: Such checking is redundant, it is enough to simply do Linq query with where clause, who cares if user asks for non existant category

Comment (Buztas) [reply Shopipy.ServiceManagement/Controllers/ServiceController.cs:53]: the app is definitely not gonna crash

**Review by suetology — COMMENTED** [review]
It's very good Justas mldc approved
image

**Review by piatrashkakanstantinass — DISMISSED** [review]
Definitely based sigma boy

File: Shopipy.ApiService/Middleware/BusinessExistsMiddleware.cs [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+using Shopipy.BusinessManagement.Services;
+
+namespace Shopipy.Persistence.Data.Middleware
+{
+    public class BusinessExistsMiddleware
```
Comment (suetology) [inline Shopipy.ApiService/Middleware/BusinessExistsMiddleware.cs:6]: I this this name is more appropriate:
```suggestion
    public class SigmaClass
```

Comment (Buztas) [reply Shopipy.ApiService/Middleware/BusinessExistsMiddleware.cs]: Uhm, this is not correct namington.

Comment (piatrashkakanstantinass) [reply Shopipy.ApiService/Middleware/BusinessExistsMiddleware.cs]: @suetology I definitely agree with the naming, our app is crashing now :(

File: Shopipy.ApiService/Program.cs [hunk 13]
```diff
@@ -23,6 +31,7 @@
 
 // Add services to the container.
 builder.Services.AddProblemDetails();
+builder.Services.AddScoped<BusinessService>();
```
[the remark this answers is not in the document]
Comment (piatrashkakanstantinass) [reply Shopipy.ApiService/Program.cs:34]: Then something is wrong as BusinessService is already injected.

File: Shopipy.AppointmentManagement/Services/AppointmentService.cs (hunk 5 above)
Comment (Buztas) [reply Shopipy.AppointmentManagement/Services/AppointmentService.cs:19]: userService only has one method ( getCurrentUserAsync ), I'll wait for the PR to be merged : |

**Review by imbieras — DISMISSED** [review]
LGTM

File: Shopipy.ServiceManagement/Controllers/ServiceController.cs [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+[ApiController]
+[Route("[controller]/{businessId}")]
+public class ServiceController(
+    IServiceManagementService serviceManagementService, 
+    ICategoryService categoryService, 
+    IMapper mapper
+) : ControllerBase
+{
+    
+    [HttpGet("services")]
```
Comment (piatrashkakanstantinass) [inline Shopipy.ServiceManagement/Controllers/ServiceController.cs:19]: You can extract services as well and leave it as `[HttpGet]`

**Review by piatrashkakanstantinass — APPROVED** [review]

File: Shopipy.AppointmentManagement/DTOs/AppointmentRequestDto.cs [hunk 15]
```diff
@@ -0,0 +1,22 @@
+using System.ComponentModel.DataAnnotations;
```
Comment (piatrashkakanstantinass) [inline Shopipy.AppointmentManagement/DTOs/AppointmentRequestDto.cs:1]: Why do you have appointment package if I see appointment logic in services already?

**Comment by suetology** [comment]
sumerginkit jau, nuo trečiadienio kaba👿👿👿

**Pull request merged**